pkg/apis/flowcontrol/types.go

```diff
@@ -0,0 +1,278 @@
+	PriorityLevelConfigurationConditionFalse   PriorityLevelConfigurationConditionStatus = "False"
+	PriorityLevelConfigurationConditionUnknown PriorityLevelConfigurationConditionStatus = "Unknown"
+)
+
+// These are valid flow-dingtinguisher methods.
+const (
+	FlowDistinguisherMethodNone        = "None"
+	FlowDistinguisherMethodByUserName  = "ByUserName"
+	FlowDistinguisherMethodByNamespace = "ByNamespace"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchema defines the schema of a group of flows. Note that a flow makes up of a set of inbound API requests with
+// similar attributes and is identified by a pair of strings: the name of the FlowSchema and a "flow distinguisher".
+type FlowSchema struct {
+	metav1.TypeMeta
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+	// Specification of the desired behavior of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec FlowSchemaSpec
+	// Current status of a flow-schema.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status FlowSchemaStatus
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// FlowSchemaList is a list of FlowSchema objects.
+type FlowSchemaList struct {
+	metav1.TypeMeta
+	// Standard list metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta
+
+	// Items is a list of flow-schemas.
+	Items []FlowSchema
+}
+
+// FlowSchemaSpec describes how the flow-schema's specification looks like.
+type FlowSchemaSpec struct {
+	// PriorityLevelConfigurationRef can reference a PriorityLevelConfiguration in the cluster. If the reference cannot
+	// be resolved, the flow-schema will ignored and marked as invalid in its status.
+	PriorityLevelConfigurationRef PriorityLevelConfigurationRef
+	// MatchingPrecedence is used to choose among the FlowSchemas that match a given request.  The chosen
+	// FlowSchema is among those with the numerically lowest (which we take to be logically highest)
+	// MatchingPrecedence.  Each MatchingPrecedence value must be non-negative.
+	MatchingPrecedence int32
+	// DistinguisherMethod defines how to classify requests into certain flows. If not empty, the flow's identifier is computed
+	// according to specification of the distinguisher.
+	// +optional
+	DistinguisherMethod FlowDistinguisherMethod
+	// Rules holds all the RuleWithUser for this FlowSchema.
+	Rules []PolicyRuleWithSubject
+}
+
+// FlowDistinguisherMethod specifies the method of a flow distinguisher.
+type FlowDistinguisherMethod struct {
+	// Name is the name of flow distinguisher method
```

Comment:
I'd prefer "Type" instead of name. Since this will grow into a union / one-of type.

Reply:
OK with me.

Reply:
s/FlowDistinguisherMethodName/FlowDistinguisherMethodType/